### ISSUE TITLE
[ALGO] Heterogeneous ZeRO-3 shard rebalancing with VRAM-proportional distribution

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#46]
## Summary

Extend ZeRO-3 optimizer state sharding to support **heterogeneous GPU clusters** by distributing parameter shards proportional to each device's available VRAM, with dynamic rebalancing triggered on OOM events.

---

## Motivation

Current ZeRO-3 implementations (DeepSpeed, Megatron) assume homogeneous GPU memory. In mixed clusters (e.g., H100 80GB + A6000 48GB), uniform sharding wastes H100 capacity or OOMs the A6000. VRAM-proportional distribution eliminates this imbalance.

---

### VRAM-Proportional Shard Assignment

```
Given ranks R = {r0: H100 80GB, r1: A6000 48GB, r2: A6000 48GB, r3: H100 80GB}
Total VRAM = 80 + 48 + 48 + 80 = 256 GB

Shard weight w_i = VRAM_i / sum(VRAM)
  w_r0 = 80/256 = 0.3125
  w_r1 = 48/256 = 0.1875
  w_r2 = 48/256 = 0.1875
  w_r3 = 80/256 = 0.3125

For model with P total parameters:
  shard_size_ri = ceil(P * w_i)

Example: LLaMA-70B (140B fp16 params = 280 GB total)
  r0 (H100): 87.5 GB shard
  r1 (A6000): 52.5 GB shard  → exceeds 48GB! → triggers rebalance
  r2 (A6000): 52.5 GB shard  → exceeds 48GB! → triggers rebalance
  r3 (H100): 87.5 GB shard
```

### VRAM-Aware Capping

```python
def compute_shard_sizes(vram_map: dict[int, float], total_params: int,
                         safety_margin: float = 0.90) -> dict[int, int]:
    total_vram = sum(vram_map.values())
    raw_shards = {r: int(total_params * (v / total_vram))
                  for r, v in vram_map.items()}
    # Cap each shard at safety_margin * VRAM (leave room for activations)
    capped = {r: min(raw_shards[r], int(v * safety_margin / BYTES_PER_PARAM))
              for r, v in vram_map.items()}
    # Redistribute overflow to highest-VRAM ranks
    overflow = total_params - sum(capped.values())
    # ... greedy redistribution to uncapped ranks
    return capped
```

### Dynamic Rebalancing on OOM

```
Trigger: torch.cuda.OutOfMemoryError caught during forward/backward pass

Rebalancing protocol:
  1. Rank that OOM'd broadcasts OOM signal via dist.broadcast(oom_tensor, src=oom_rank)
  2. All ranks pause at next allreduce barrier
  3. OOM rank reports current VRAM usage: dist.gather(usage_tensor)
  4. Coordinator rank (rank 0) recomputes shard sizes with shrink factor:
       new_vram_budget_r = 0.85 × old_vram_budget_r  (for OOM rank)
  5. Shard migration:
       - OOM rank sends excess param chunk to least-loaded rank via send/recv
       - cudaMemcpyPeerAsync for same-node, NCCL for cross-node
  6. All ranks update local param_to_rank mapping
  7. Resume training from last checkpoint (or current step if using gradient accumulation)
```

### Shard Metadata

```python
@dataclass
class ShardMetadata:
    rank: int
    param_start_idx: int    # global flat param index
    param_end_idx: int
    vram_budget_bytes: int
    last_rebalance_step: int = -1
    oom_count: int = 0
```

---

## Integration Points

- `megatron/core/distributed/distributed_data_parallel.py` → `_reduce_scatter_grads()`
- `megatron/core/optimizer/distributed_optimizer.py` → `_build_model_and_main_param_groups()`
- New module: `neuron_sp/zero3/hetero_shard_manager.py`

---

## Benchmark Plan

- [ ] Memory utilization per rank: uniform sharding vs VRAM-proportional
- [ ] Time-to-OOM: how many steps before OOM with uniform vs proportional
- [ ] Rebalancing overhead: latency of one rebalance event (target: < 2 training steps)
- [ ] Throughput (samples/sec) on H100 × 2 + A6000 × 2 mixed cluster

---

## Acceptance Criteria

- [ ] Shard sizes proportional to VRAM within ±2% error
- [ ] Dynamic rebalancing completes without process crash
- [ ] After rebalancing, VRAM utilization stays below 90% on all ranks
- [ ] No gradient correctness regression vs single-GPU baseline (atol=1e-4 for fp32 weights)
- [ ] Works with Megatron-LM TP=1 and TP=4 configurations

---

## References

- [DeepSpeed ZeRO-3 paper](https://arxiv.org/abs/1910.02054)
- [Megatron-LM distributed optimizer](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/optimizer/distrib_optimizer.py)
- [Whale: Efficient Giant Model Training over Heterogeneous GPUs](https://arxiv.org/abs/2011.10780)
